platform-sdk/swirlds-jasperdb/src/main/java/com/swirlds/merkledb/MerkleDbCompactionCoordinator.java

```diff
@@ -0,0 +1,282 @@
+
+    /**
+     * Enables background compaction.
+     */
+    void enableBackgroundCompaction() {
+        compactionEnabled.set(true);
+    }
+
+    /**
+     * Stops all compactions in progress and disables background compaction.
+     * All subsequent calls to compacting methods will be ignored until {@link #enableBackgroundCompaction()} is called.
+     */
+    void stopAndDisableBackgroundCompaction() {
+        synchronized (compactionFuturesByName) {
+            compactionFuturesByName.forEach((k, v) -> v.cancel());
+            compactionFuturesByName.clear();
+            compactionEnabled.set(false);
+        }
+    }
+
+    /**
+     * Submits a compaction task for execution. If a compaction task for the same storage type is already in progress,
+     * the call is effectively no op.
+     * @param task a compaction task to execute
+     */
+    private void submitCompactionTaskForExecution(CompactionTask task) {
+        if (!compactionEnabled.get()) {
+            return;
+        }
+
+        ExecutorService executor = createOrGetCompactingExecutor();
+
+        synchronized (compactionFuturesByName) {
+            if (compactionFuturesByName.containsKey(task.id)) {
+                CompletableFuture<?> completableFuture =
+                        compactionFuturesByName.get(task.id).asCompletableFuture();
+                if (completableFuture.isDone()) {
+                    compactionFuturesByName.remove(task.id);
+                } else {
+                    logger.debug(MERKLE_DB.getMarker(), "Compaction for {} is already in progress", task.id);
+                    return;
+                }
+            }
+
+            InterruptibleCompletableFuture<Boolean> future =
+                    InterruptibleCompletableFuture.runAsyncInterruptibly(task, executor);
+            future.asCompletableFuture().thenAccept(result -> {
+                if (result) {
+                    statisticsUpdater.updateStoreFileStats();
+                    statisticsUpdater.updateOffHeapStats();
+                    logger.info("Finished compaction for " + task.id);
+                }
+            });
+            compactionFuturesByName.put(task.id, future);
+        }
+    }
+
+    /**
+     * Creates a thread pool for compaction tasks if it does not exist yet, or returns the existing one.
+     * If the existing thread pool is shutdown, it will be replaced with a new one. Note that there is only one
+     * compacting thread-pool per JVM.
+     * @return a thread pool for compaction tasks
+     */
+    ExecutorService createOrGetCompactingExecutor() {
+        ExecutorService executorService = compactionExecutorServiceRef.get();
+        if (executorService == null) {
+            executorService = new ThreadPoolExecutor(
+                    getCompactingThreadNumber(),
+                    getCompactingThreadNumber(),
+                    0L,
+                    TimeUnit.MILLISECONDS,
+                    new LinkedBlockingDeque<>(),
```

Comment:
This should be `LinkedBlockingQueue`.  I don't think we gain anything from the additional overhead of a `Deque`.

Reply:
Fixed